pycroft/lib/mpsk_client.py

```diff
@@ -0,0 +1,81 @@
+#  Copyright (c) 2024. The Pycroft Authors. See the AUTHORS file.
+#  This file is part of the Pycroft project and licensed under the terms of
+#  the Apache License, Version 2.0. See the LICENSE file for details
+from pycroft.model.mpsk_client import MPSKClient
+from pycroft.model.types import AmountExceededError
+from pycroft.model.user import User
+from pycroft.model.session import session
+from pycroft.lib.logging import log_user_event
+from pycroft.helpers.i18n import deferred_gettext
+
+
+def mpsk_delete(host: MPSKClient, processor: User) -> None:
+    message = deferred_gettext("Deleted host '{}'.").format(host.name)
+    log_user_event(author=processor, user=host.owner, message=message.to_json())
+
+    session.delete(host)
+
+
+def change_mac(client: MPSKClient, mac: str, processor: User) -> MPSKClient:
+    """
+    This method will change the mac address of the given interface to the new
+    mac address.
+
+    :param interface: the interface which should become a new mac address.
+    :param mac: the new mac address.
+    :param processor: the user who initiated the mac address change.
+    :return: the changed interface with the new mac address.
+    """
+    old_mac = client.mac
+    client.mac = mac
+    message = deferred_gettext("Changed MAC address from {} to {}.").format(old_mac, mac)
+    if client.owner:
+        log_user_event(message.to_json(), processor, client.owner)
+    return client
+
+
+def mpsk_client_create(owner: User, name: str, mac: str, processor: User) -> MPSKClient:
+
+    if len(owner.mpsks) >= 10:
+        raise AmountExceededError(
+            "the limit of added mpsks clients is exceeded", limit=10, actual=len(owner.mpsks)
+        )
```

Comment:
Contrary as suggested by me in https://github.com/agdsn/pycroft/pull/749#issuecomment-2360248583, you introduced the limit in the lib.
This also prevents admins to add more devices.

Reply:
This could also be solved by adding an extra parameter to the function that disables the check if needed.

Reply:
Yeah but then I can not test that also I dont think any one would use more then 10 but I can do so

Reply:
I agree with @ibot3, it is better to have flexibility for the admins. Don't worry about testing it, adding a test for the API can be done later (I will do it).

Reply:
should be In :)

Reply:
I can only speak for me, but I have more than 10 devices at home that only can use PSK wifi.
Especially when using smart-home stuff, this can add-up quite fast.

Reply:
thats true but I dont think that these members are not using a router also maybe its a nice way to get active members :)

Reply:
But I can also set it higher what do you think would be a good number of devices? 50 Sounds a bit much for me in my opinion

Reply:
If the wifi network works as planned in the future, then we don't want that users have their own wifi router, we might even want to prohibit other wifi networks than ours.
However, this is not the case yet.
But our wifi should work in a way that the users don't have the need for another wifi-router.

I would suggest increasing the limit to 30.

Reply:
true I now set it to 30 thank

Reply:
> This could also be solved by adding an extra parameter to the function that disables the check if needed.

Sorry for the confusion, but I just now saw that @ibot3 suggested adding a boolean flag. I was agreeing only with his [first message](https://github.com/agdsn/pycroft/pull/749#discussion_r1766468448).

As a rule of thumb, **boolean flags are a [code smell](https://martinfowler.com/bliki/FlagArgument.html)**.

Think about it, this flag makes the problem worse: Instead of removing the check from the lib function, it leaves it in, but _conditionally_. This makes `mpks_client_create` more complex (e.g., it duplicates the test efforts).
This kind of rate limit is not part of `pycroft`s business logic, it's an implementation decision from the API.  Since the API should depend on the lib, but not vice versa, just the fact that the word `api` is now part of a lib function is a clear sign that we're entering the depths of ~~hell~~ accidental complexity.

1. The API endpoint should check it (when we implement it).
2. I also agree with 30, but the fact that we are arguing about the value of a hard-coded number shows that it should be a *configurable environment variable*.

So for now I would ask you to just remove the check, and we can agree on a pre-configured limit of 30 when the API gets implemented.